provider/adapters/src/fork/by_key.rs

```diff
@@ -162,6 +162,21 @@ where
     }
 }
 
+// TODO(#1893): Instead of going through DynProvider, this should only implement ResourceProvider
+// for the ResourceProviders that P0 and P1 implement. I haven't figured out a way to do this though
+// as we'll need two blanket implementations that could conflict.
+#[cfg(feature = "datagen")] // this is not production-ready
+impl<M, P0, P1> ResourceProvider<M> for ForkByKeyProvider<P0, P1>
+where
```

Comment:
This is fine and is the way I would expect this to be done. However, a more general solution would be something like `.as_resource()` that converts *any* `DynProvider` to a compatible `ResourceProvider`. I'm open to having it implemented directly on ForkByKeyProvider, but it should just wrap over the general function.

Reply:
Also, the use case here is very narrow. There are not many cases where we want to convert from `DynProvider` to `ResourceProvider`.

Reply:
> There are not many cases where we want to convert from DynProvider to ResourceProvider.

That's not what I'm trying to do. The scenario here is that you have `A: ResourceProvider<Foo>` and `B: ResourceProvider<Bar>`, but you need `C: ResourceProvider<Foo> + ResourceProvider<Bar>`. If `A` and `B` implement `DynProvider<Foo>` and `DynProvider<Bar>`, respectively, this hack works. But it will also make `C` implement `ResourceProvider<Batz>` which will fail with a key error, something that should not happen with a `ResourceProvider`.

Reply:
So, to be clear, when you want to load `Bar`, what this is doing is it calls `A::load_payload`, which calls either `A::as_deserializing()` or `A::as_downcasting()` (suppose deserializing), which calls `A::load_buffer`, which may or may not find the buffer. If it doesn't find the buffer, then we try the whole stack again with `B`.

OK, this seems reasonable.

I think it's probably worth making it general?

Reply:
> So, to be clear, when you want to load Bar, what this is doing is it calls A::load_payload

Not all `ResourceProvider`s are backed by `DynProvider`s, crabbake in particular is not.

Reply:
To clarify, the advantage of not calling `A::load_payload` just to check if it supports the key is that we can do DCE if we statically know which resources `A` supports. For `BufferProvider` and `AnyProvider` it's always all of them because they have blanket implementations, but that is not generally true for other providers.